legion/engine/physics/mesh_splitter_utils/mesh_splitter.hpp

```diff
@@ -0,0 +1,749 @@
+#pragma once
+
+#include <core/core.hpp>
+#include <physics/mesh_splitter_utils/mesh_splitter_typedefs.hpp>
+#include <physics/mesh_splitter_utils/mesh_half_edge.hpp>
+#include <physics/mesh_splitter_utils/splittable_polygon.h>
+#include <rendering/components/renderable.hpp>
+#include <physics/mesh_splitter_utils/primitive_mesh.h>
+#include <rendering/components/renderable.hpp>
+#include <physics/mesh_splitter_utils/half_edge_finder.h>
+#include <physics/mesh_splitter_utils/mesh_splitter_debug_helpers.h>
+#include <physics/mesh_splitter_utils/intersecting_polygon_organizer.hpp>
+#include <physics/mesh_splitter_utils/mesh_split_params.hpp>
+#include <physics/mesh_splitter_utils/intersection_edge_info.h>
+
+namespace legion::physics
+{
+    struct MeshSplitter
+    {
+
+        ecs::entity_handle owner;
+        std::vector<ecs::entity_handle> splitTester;
+
+        rendering::material_handle ownerMaterialH;
+
+        std::vector<SplittablePolygonPtr> meshPolygons;
+
+        MeshSplitterDebugHelper debugHelper;
+
+        //------------------------------------------------------ Function related to Mesh Splitter Initialization ---------------------------------------------//
+
+        /** @brief Creates a Half-Edge Data structure around the mesh and
+       * @param entity the entity that this MeshSplitter is attached to
+       */
+        void InitializePolygons(ecs::entity_handle entity)
+        {
+            owner = entity;
+
+            auto [meshFilter, meshRenderer] = entity.get_component_handles<rendering::mesh_renderable>();
+
+            ownerMaterialH = meshRenderer.read().material;
+
+            auto [posH, rotH, scaleH] = entity.get_component_handles<transform>();
+
+            if (meshFilter && posH && rotH && scaleH)
+            {
+                log::debug("Mesh and Transform found");
+                std::queue<meshHalfEdgePtr> meshHalfEdges;
+
+                //auto renderable = renderable.read();
+                mesh& mesh = meshFilter.read().get().second;
+
+                const math::mat4 transform = math::compose(scaleH.read(), rotH.read(), posH.read());
+
+                HalfEdgeFinder edgeFinder;
+                edgeFinder.FindHalfEdge(mesh, transform, meshHalfEdges);
+
+                BFSPolygonize(meshHalfEdges, transform);
+
+                log::debug("Mesh vertices {}, Mesh indices {}", mesh.vertices.size(), mesh.indices.size());
+
+            }
+            else
+            {
+                log::warn("The given entity does not have a meshHandle!");
+            }
+
+        }
+
+        /** @brief Given a queue of edges and a transform,
+        * populates the std::vector 'meshPolygons' using BFS
+        */
+        void BFSPolygonize(std::queue<meshHalfEdgePtr>& halfEdgeQueue, const math::mat4& transform)
```

Comment:
You should probably comment that halfEdgeQueue will be emptied, and will not contain any data after this function